12 user can create an account

amandahamynen opened the pull request:
User can create an account with a username, password and an email address. The email address has to follow the standard form of emails, and there can be only one user with the same email-address. 

Closes #12

VSirvio reviewed (changes requested):
I think your code is very high quality and there isn't really much to fix. I have just a couple of points about usability and security that you could check.

**Usability:**
- Maybe it's not necessary to confirm the user if they are sure they want to register
- When registration fails, you could maybe show the exact error message from the backend instead of the rather general "Registering failed"
- It would be good to have the HTML attribute 'type="email"' in the email input element, so also the browser would do some validation for the email address

**Security:**
- We could probably leave out the user listing endpoint in the backend at this time, because such a user listing should normally be visible only for admins, and we don't yet have any way to find out whether a person is an admin or not
- It would maybe be better not to return the created user info from the backend when registering, because it has sensitive data included (especially the password hash) and we aren't actually using the returned data to anything
- Another point that would probably be too much work to implement right now, but is worth thinking about, is that it would probably be a good idea to hash the password already in the frontend, so that the actual password would never be sent through the network but just the hash

All in all, I think you did very good job with this user story!

amandahamynen commented:
> I think your code is very high quality and there isn't really much to fix. I have just a couple of points about usability and security that you could check.
> 
> **Usability:**
> 
> * Maybe it's not necessary to confirm the user if they are sure they want to register
> * When registration fails, you could maybe show the exact error message from the backend instead of the rather general "Registering failed"
> * It would be good to have the HTML attribute 'type="email"' in the email input element, so also the browser would do some validation for the email address
> 
> **Security:**
> 
> * We could probably leave out the user listing endpoint in the backend at this time, because such a user listing should normally be visible only for admins, and we don't yet have any way to find out whether a person is an admin or not
> * It would maybe be better not to return the created user info from the backend when registering, because it has sensitive data included (especially the password hash) and we aren't actually using the returned data to anything
> * (Edit: Actually forget about this. This can't be done with salted passwords.) Another point that would probably be too much work to implement right now, but is worth thinking about, is that it would probably be a good idea to hash the password already in the frontend, so that the actual password would never be sent through the network but just the hash
> 
> All in all, I think you did very good job with this user story!

Usability:
- Removed the confirmation when the user registers
- Error messages fixed
- Added type="email” for the email-input. Now the user cannot submit the registeration unless it passes basic validation in addition to the rules set backend.

Security:
- Removed user listing from backend
- Removed returning user info from backend when registering
- Allright, thank you for informing me about that! 

Thank you so much for a very thoughtful and thorough review once again, @VSirvio ! :)
